src/pages/resultsView/survival/SurvivalUtil.tsx

```diff
@@ -152,4 +184,87 @@ export function convertScatterDataToDownloadData(patientData: any[]): any[] {
     })
 
     return downloadData;
+}
+
+/**
+ * Down sampling based on the hypotenuse difference.
+ * The cross dot is guaranteed to be drawn when hidden dots are taking into account.
+ * @param {DownSampling} opts
+ * @returns {ScatterData[]}
+ */
+export function downSampling(data: ScatterData[], opts: DownSamplingOpts): ScatterData[] {
+    let x = data.map(item => item.x);
+    let y = data.map(item => item.y);
+
+    let xMax = _.max(x) || 0;
+    let xMin = _.min(x) || 0;
+
+    let yMax = _.max(y) || 0;
+    let yMin = _.min(y) || 0;
+
+    if (opts.xDenominator <= 0 || opts.yDenominator <= 0)
+        return data;
+
+    let timeThreshold = (xMax - xMin) / opts.xDenominator;
+    let survivalRateThreshold = (yMax - yMin) / opts.yDenominator;
+    let averageThreshold = Math.hypot(timeThreshold, survivalRateThreshold);
+    let lastVisibleDataPoint = {
+        x: 0,
+        y: 0
+    };
+    let lastDataPoint = {
+        x: 0,
+        y: 0
+    };
+
+    return data.filter(function (dataItem, index) {
+        let isVisibleDot = _.isUndefined(dataItem.opacity) || dataItem.opacity > 0;
+        if (index == 0) {
+            lastDataPoint.x = dataItem.x;
+            lastDataPoint.y = dataItem.y;
+            lastVisibleDataPoint.x = dataItem.x;
+            lastVisibleDataPoint.y = dataItem.y;
+            return true;
+        }
+        let distance = 0;
+        if (isVisibleDot) {
+            distance = Math.hypot(dataItem.x - lastVisibleDataPoint.x, dataItem.y - lastVisibleDataPoint.y)
+        } else {
+            distance = Math.hypot(dataItem.x - lastDataPoint.x, dataItem.y - lastDataPoint.y);
+        }
+        if (distance > averageThreshold) {
+            lastDataPoint.x = dataItem.x;
+            lastDataPoint.y = dataItem.y;
+            if (isVisibleDot) {
+                lastVisibleDataPoint.x = dataItem.x;
+                lastVisibleDataPoint.y = dataItem.y;
+            }
+            return true;
+        } else {
+            return false;
+        }
+    });
+}
+
+export function filteringScatterData(allScatterData: GroupedScatterData, filters: SurvivalPlotFilters | undefined, downSamplingOpts: DownSamplingOpts):GroupedScatterData {
+    let filteredData = _.cloneDeep(allScatterData);
+    _.forEach(filteredData, (value:SurvivalCurveData, key:string) => {
+        if (value.numOfCases > downSamplingOpts.threshold) {
+            if (filters) {
+                let filteredIndex: number[] = [];
+                _.forEach(value.scatter, (_val, _key) => {
```

Comment:
in places where we are probably dealing with large amounts of data, we should use the native api instead of lodash.  you can use arr.forEach, arr.filter and even arr.includes.    (Arr.includes might have typing issue)